memorystore/valkey/caching/GUIDE.md

```diff
@@ -0,0 +1,189 @@
+- **PostgreSQL Database:** Your source of truth for all data. The cache reduces how often the app queries this database.
+- **Google Cloud Infrastructure:** Deployed using Terraform, you can host the application on Cloud Run, store data in Cloud SQL for PostgreSQL, and leverage Memorystore for Valkey.
+
+## Step-by-step Guide
+
+To begin, we are will generate an API with the following routes:
+
+_create_: For creating new items.
+_delete_: For creating deleting items.
+_create_: For creating new items.
+_create_: For creating new items.
+
+### Creating a new application
+
+The first step is to initialize a brand new Spring Boot application. The [offical guide](https://spring.io/guides/gs/spring-boot) demonstrates how to generate a new project using [Spring Initializer](https://start.spring.io/).
+
+1. We chosen `Maven` as the project type for this demonstration..
+2. Select Sprint Boot version 3.4.1
+3. Complete the appropriate metadata.
+4. Choose your preffered Packing for downloading
+5. Select `Java 17` for your Java version.
+6. Finally, generate and extract the files.
+
+### Installing additional dependencies
+
+Next, ensure the following dependencies have been added to your POM.xml file.
+
+#### Jedis
+
+Add the folowing snippet toconnect directly to the Memorystore for Valkey instance.
+
+```xml
+        <!-- Jedis: Redis Java Client -->
+        <dependency>
+            <groupId>redis.clients</groupId>
+            <artifactId>jedis</artifactId>
+            <version>4.3.0</version> <!-- Use the latest version -->
+        </dependency>
+```
+
+#### Jakarta
+
+To ensure that our api routes are correctly validated. Add the following dependency:
+
+```xml
+   <!-- Add Validation support-->
+   <dependency>
+      <groupId>jakarta.validation</groupId>
+      <artifactId>jakarta.validation-api</artifactId>
+      <version>3.0.2</version>
+   </dependency>
+```
+
+### Connecting our Service layer
+
+Next, add the following to route logic to the API.
+
+### Writing to the cache
+
+Adding an entry will add the new item to the database. Once created, the id return from the database will be updated to a new object with the entries attributes. This new object will be added to the Memorystore cach with the appropirate Time-to-live (TTL) value.
+
+```java
+
+/** Import the Jedis library */
+import redis.clients.jedis.Jedis;
+
+/** Creating an item with Time-to-live (TTL) */
+public long create(Item item) {
+   /** Save the item in the database */
+   long itemId = itemsRepository.create(item);
+
+   /** Create a new object with the saved database id */
+   Item createdItem = new Item(
+   itemId,
+   item.getName(),
+   item.getDescription(),
+   item.getPrice()
+   );
+
+   /** Cache the data in Memorystore for valkey with the Time-to-live value **/
+   String idString = Long.toString(itemId);
+   jedis.set(idString, createdItem.toJSONObject().toString());
```

Comment:
expire can be included in set() as a SetParam https://github.com/redis/jedis/blob/3e537f948d17a5511db3b8f57cd4e093407852b0/src/main/java/redis/clients/jedis/commands/StringBinaryCommands.java#L14

This is a more recommended pattern as it cuts QPS in half versues calling set() followed by expire() every time.

Reply:
recommend use this throughout.